### PR TITLE
Extend feature

### PR DESCRIPTION
**Pull request by vkz**
extend() feature requested in #22

### CODE REVIEW

**Comment by vkz** [comment]
cc @indutny

File: lib/bemhtml/compiler.js [hunk 1]
```diff
@@ -375,14 +397,57 @@ Compiler.prototype.flatten = function flatten(expr) {
...
   }, this);
 };
 
-Compiler.prototype.replaceBody = function replaceBody(body) {
+Compiler.prototype.replaceBody = function replaceBody(body, extendedModeDetected) {
   var self = this;
 
+  // Desugar bodies of replace() and extend()
+  if (extendedModeDetected) {
+    body.body = estraverse.replace(body.body, {
```
Comment (indutny) [inline lib/bemhtml/compiler.js]: Please move it to it's own method.

File: lib/bemhtml/compiler.js [hunk 2]
```diff
@@ -134,7 +146,15 @@ Compiler.prototype.markMatches = function markMatches(ast) {
     };
   }
 
-  if (ast.type !== 'CallExpression') return ast;
+  if (ast.type !== 'CallExpression') {
+    if (ast.type === 'Identifier' &&
+        (ast.name === 'replace' || ast.name === 'extend')) {
+      // Desugar`replace' mode predicate and set the flag 
+      ast.extendedModeDetected = ast.name;
```
Comment (indutny) [inline lib/bemhtml/compiler.js]: Name `extendedModeDetected` suggests that the value is boolean, but in fact it is a mode itself. Why not just call it `extendedMode` then?

File: lib/bemhtml/compiler.js [hunk 3]
```diff
@@ -375,14 +397,57 @@ Compiler.prototype.flatten = function flatten(expr) {
...
         arguments: t.predicates
       },
-      arguments: [ this.replaceBody(t.body) ]
+      arguments: [ this.replaceBody(t.body, extendedModeDetected) ]
     };
   }, this);
 };
 
-Compiler.prototype.replaceBody = function replaceBody(body) {
+Compiler.prototype.replaceBody = function replaceBody(body, extendedModeDetected) {
```
Comment (indutny) [inline lib/bemhtml/compiler.js]: 80 column limit everywhere.

File: lib/bemhtml/compiler.js [hunk 4]
```diff
@@ -375,14 +397,57 @@ Compiler.prototype.flatten = function flatten(expr) {
...
+        }
+      },
+      leave: function (ast) {
+        if (ast.type === 'ReturnStatement') {
+          var applyCtxArgNode,
+              applyCtxNode;
+          if (extendedModeDetected === 'replace') {
+            applyCtxArgNode = ast.argument;
+          } else if (extendedModeDetected === 'extend') {
+            applyCtxArgNode = {
```
Comment (indutny) [inline lib/bemhtml/compiler.js]: Style! :) Missing semicolon, odd braces alignment. Please keep it the same as it is in the rest of the file.

File: lib/bemhtml/runtime.js [hunk 5]
```diff
@@ -127,6 +129,14 @@ module.exports = function() {
   function bem() { return mode.call(this, 'bem'); };
   function mix() { return mode.call(this, 'mix'); };
   function content() { return mode.call(this, 'content'); };
+  function replace() {
+    var self = this;
+    return function (body) { return self.def()(function () { return applyCtx(body()) }) };
```
Comment (indutny) [inline lib/bemhtml/runtime.js]: 80 column limit ;)

Comment (indutny) [inline lib/bemhtml/runtime.js]: And a missing semicolon. Perhaps we should use jshint in `npm test`?

File: test/api-test.js [hunk 6]
```diff
@@ -22,6 +22,19 @@ describe('BEMHTML compiler', function() {
     });
   }
 
+  function testDesugaring(fn, fnDesugared, options) {
+    if (!options) options = {};
+    var ibem = options.baseTmpl !== false ? require('./fixtures/i-bem.bemhtml') : '';
```
Comment (indutny) [inline test/api-test.js]: 80 column limit, most likely. Also could be:

```javascript
var ibem = '';
if (options.baseTmpl !== false) ibem = ...
```

File: test/api-test.js [hunk 7]
```diff
@@ -347,8 +360,44 @@ describe('BEMHTML compiler', function() {
...
+      block('b1').content()('ok');
+      block('b2').content()('replaced');
+      block('b1').replace()(function () {return {block: 'b2'}});
+    }, {block: 'b1'}, '<div class="b2">replaced</div>')
+  })
+
+  it('should desugar extend() mode', function () {
+    testDesugaring(function () {
+      block('b1').extend()(function() { return {elem: 'e'}; });
+      block('b1').elem('e').extend()(function() { return {"mods" : {"pseudo" : "yes"}}; });
```
Comment (indutny) [inline test/api-test.js]: 80 column limit.

**Comment by indutny** [comment]
Do you mind if we'll pull both features in this PR? It doesn't seem that #22 will have any relevance after landing this.

**Comment by vkz** [comment]
@indutny both features are now in this PR with clean history. Can u pls have another look.

File: lib/bemhtml/compiler.js [hunk 8]
```diff
@@ -375,14 +398,68 @@ Compiler.prototype.flatten = function flatten(expr) {
...
+                ast.argument
+              ]
+            };
+          }
+
+          applyCtxNode = {
+              type: 'CallExpression',
+              callee: { type: 'Identifier', name: 'applyCtx'},
+              arguments: [ applyCtxArgNode ]
+            };
```
Comment (indutny) [inline lib/bemhtml/compiler.js]: Invalid indent here and above.

**Comment by indutny** [comment]
One very last style nit, otherwise LGTM :)

**Comment by vkz** [comment]
argh, this style thing is killing me :) I added whitespace and line-length autodetection to my dev tools. Will have a go at style hinting later. Thx for catching

**Comment by indutny** [comment]
Landed and released in 1.0.0 . Thank you!

**Pull request closed without merging**